### PR TITLE
FRR package enhancements

### PR DESCRIPTION
**Pull request by bmhughes**
Started off tidying up the BFD integrating in #835 and found a few other things to tidy up.

1. Extend #835 to better integrate with the current FRR configuration style and add support for BFD to BGP, OSPF and OSPFv3 daemons.
2. Make the OSPF and OSPFv3 configuration flow more alike.
3. Allow the OSPF/OSPFv3 default area to be unspecified as long as the area is specified on the interface for situations where a default is not preferable.
3. Add a check to prevent a situation where BGP can be started with no local AS set.
4. Tidy up the configuration by only generating ACCEPTFILTER when the convenience filter functions in the UI have been used.
5. Add agentx support to FRR to allow SNMP data retrieval when used with the net-snmp package.

Tested on 2.4.5 release with FRR 7.2.1 and 7.3.

### CODE REVIEW

**Comment by bmhughes** [comment]
Bugtracker issue: https://redmine.pfsense.org/issues/10619

**Comment by jim-p** [comment]
I see you are still making some changes, is this ready for review or are there still more changes pending?

**Comment by bmhughes** [comment]
I just realised I'd missed the out the FRR rc file to load the SNMP modules. I'm just testing it now but that should be it.

File: net/frr7/Makefile [hunk 1]
```diff
@@ -34,7 +34,7 @@ INFO=		frr
 
 OPTIONS_DEFINE=		CONFIGROLLBACKS FPM DATACENTER LDPD PAM SHELLACCESS \
 			TCPSOCKETS VTYSH MANPAGES MULTIPATH TCMALLOC RPKI SNMP
-OPTIONS_DEFAULT=	VTYSH RPKI FPM MANPAGES
+OPTIONS_DEFAULT=	VTYSH RPKI FPM MANPAGES SNMP
```
Comment (rbgarga) [inline net/frr7/Makefile:37]: We don't change items from port that can be changed with environment variables.  Please remove this change and submit a separate PR to pfSense main repository enabling this option on make.conf at
https://github.com/pfsense/pfsense/blob/master/tools/conf/pfPorts/make.conf#L227

**Comment by rbgarga** [comment]
@bmhughes please rebase your fork and resolve conflicts.  Other PRs were merged and ended up changing the same files

File: net/frr7/Makefile (hunk 1 above)
Comment (bmhughes) [reply net/frr7/Makefile:37]: Sure, no problem.

Comment (bmhughes) [reply net/frr7/Makefile:37]: PR: https://github.com/pfsense/pfsense/pull/4359

**Comment by bmhughes** [comment]
Done

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr.inc [hunk 2]
```diff
@@ -273,6 +274,16 @@ function frr_generate_config_rcfile() {
...
 
+	/* Get FRR daemon modules to append to daemon commands */
+	$zebradmodules = '';
+	$bgpdmodules = '';
+	$ospfdmodules = '';
+	$ospf6dmodules = '';
+	if ($config['installedpackages']['frr']['config'][0]['enablesnmpagentx']) { $zebradmodules .= '-M snmp'; }
+	if ($config['installedpackages']['frrbgp']['config'][0]['enablesnmpagentx']) { $bgpdmodules .= '-M snmp'; }
+	if ($config['installedpackages']['frrospfd']['config'][0]['enablesnmpagentx']) { $ospfdmodules .= '-M snmp'; }
+	if ($config['installedpackages']['frrospf6d']['config'][0]['enablesnmpagentx']) { $ospf6dmodules .= '-M snmp'; }
```
Comment (jim-p) [inline net/pfSense-pkg-frr/files/usr/local/pkg/frr.inc:285]: These should either be done ternary style or ["K&R, BSD KNF variant style" formatting](https://www.netgate.com/docs/pfsense/development/developer-style-guide.html#indent-style), the way they are now is not a format we want in the code.

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd.xml [hunk 3]
```diff
@@ -35,10 +35,14 @@
...
 		</tab>
+		<tab>
+			<text>Peers</text>
+			<url>pkg.php?xml=frr/frr_bfd_peers.xml</url>
+		</tab>	
 		<tab>
 			<text>[Global Settings]</text>
 			<url>pkg_edit.php?xml=frr.xml</url>
-		</tab>
+		</tab>
```
Comment (jim-p) [inline net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd.xml:45]: Extra unnecessary whitespace at the end of this line

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd_peers.xml [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+			</description>
+			<size>50</size>
+			<type>input</type>
+		</field>
+		<field>
+			<fielddescr>Multihop</fielddescr>
+			<fieldname>multihop</fieldname>
+			<description>
+				<![CDATA[
+				 Tell the BFD daemon that we should expect packets with TTL less than 254 (because it will take more than one hop) and to listen on the multihop port (4784).
```
Comment (jim-p) [inline net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd_peers.xml:118]: Reword so it does not use "we".

"Tell the BFD daemon that we should expect" could be shortened to "Expect" with no loss of meaning.

Also try to reword to eliminate the parenthetical phrase.

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd_peers.xml [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+			</description>
+			<type>checkbox</type>
+		</field>
+		<field>
+			<fielddescr>Shutdown</fielddescr>
+			<fieldname>shutdown</fieldname>
+			<description>
+				<![CDATA[
+				Enables or disables the peer.
+				When the peer is disabled an ‘administrative down’ message is sent to the remote peer.
```
Comment (jim-p) [inline net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd_peers.xml:130]: Change the quote marks to traditional double quotes, not unicode single quotes.

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd_peers.xml [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+		</field>
+
+		<field>
+			<name>Source Address/Interface</name>
+			<type>listtopic</type>
+		</field>
+		<field>
+			<fielddescr>Interface</fielddescr>
+			<fieldname>interface</fieldname>
+			<description>Select which interface we should use</description>
```
Comment (jim-p) [inline net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd_peers.xml:143]: Reword to avoid "we", suggest: "Select interfaces for use with BFD" or similar.

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd_peers.xml [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+			<source_name>name</source_name>
+			<source_value>value</source_value>
+		</field>
+		<field>
+			<fielddescr>Local Source Address</fielddescr>
+			<fieldname>local_address</fieldname>
+			<description>
+				<![CDATA[
+				Provide a local address that we should bind our peer listener to and the address we should use to send the packets.<br />
+				This option is mandatory for IPv6.
```
Comment (jim-p) [inline net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd_peers.xml:155]: Reword to avoid "we".

Suggested: "Local IP address to which BFD will bind for sending and receiving packets to peers."

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd_peers.xml [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+			</description>
+			<type>input</type>
+			<size>20</size>
+		</field>
+		<field>
+			<fielddescr>Receive Interval</fielddescr>
+			<fieldname>receiveinterval</fieldname>
+			<description>
+				<![CDATA[
+				Configures the minimum interval that this system is capable of receiving control packets.
```
Comment (jim-p) [inline net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd_peers.xml:188]: Sentence seems a bit awkward. Perhaps the middle should be "interval at which this system"

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd_peers.xml [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+				The default value is 300 milliseconds.
+				]]>
+			</description>
+			<type>input</type>
+			<size>20</size>
+		</field>
+		<field>
+			<fielddescr>Transmit Interval</fielddescr>
+			<fieldname>transmitinterval</fieldname>
+			<description>The minimum transmission interval (less jitter) that this system wants to use to send BFD control packets.</description>
```
Comment (jim-p) [inline net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd_peers.xml:199]: This reads a bit awkwardly as well.

Perhaps "The minimum transmission interval at which BFD wants to send control packets. Lower values result in less jitter."

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd_peers.xml [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+			<type>input</type>
+			<size>20</size>
+		</field>
+		<field>
+			<fielddescr>Echo Interval</fielddescr>
+			<fieldname>echointerval</fieldname>
+			<description>Configures the minimal echo receive transmission interval that this system is capable of handling.</description>
```
Comment (jim-p) [inline net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd_peers.xml:206]: Reword similar to above. Seems unnecessarily long/wordy.

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd_peers.xml [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+			<size>20</size>
+		</field>
+		<field>
+			<fielddescr>Echo Mode</fielddescr>
+			<fieldname>echo_mode</fieldname>
+			<description>
+				<![CDATA[
+				Enables or disables the echo transmission mode. This mode is disabled by default.
+				<br />
+				It is recommended that the transmission interval of control packets to be increased after enabling echo-mode to reduce bandwidth usage. For example: transmission-interval 2000.
```
Comment (jim-p) [inline net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd_peers.xml:217]: Avoid the passive voice if possible. Use phrases like "The best practice is ..."  rather than "It is recommended" (who recommends it?)

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bgp_neighbors.xml [hunk 12]
```diff
@@ -375,6 +375,48 @@
...
+		<field>
+			<name>BFD</name>
+			<type>listtopic</type>
+		</field>
+		<field>
+			<fielddescr></fielddescr>
+			<description>
+				<![CDATA[
+				<b>BFD Enable</b> : Listen for BFD events registered on the same target as this BGP neighbor.
+				When BFD peer goes down it immediately asks BGP to shutdown the connection with its neighbor and, when it goes back up, notify BGP to try to connect to it.
```
Comment (jim-p) [inline net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bgp_neighbors.xml:387]: This sentence needs reworded. It's missing words/has, awkward syntax, and seems like it could be shortened significantly.

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bgp_neighbors.xml [hunk 13]
```diff
@@ -375,6 +375,48 @@
...
+				<br></br>
+				<b>BFD Check Control Plane Failure</b>: Allow to write CBIT independence in BFD outgoing packets. Also allow to read both C-BIT value of BFD and lookup BGP peer status.
+				This command is useful when a BFD down event is caught, while the BGP peer requested that local BGP keeps the remote BGP entries as staled if such issue is detected.
```
Comment (jim-p) [inline net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bgp_neighbors.xml:390]: Should "staled" be "stale" or "stalled"?

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bgp_neighbors.xml [hunk 14]
```diff
@@ -375,6 +375,48 @@
...
+				This is the case when graceful restart is enabled, and it is wished to ignore the BD event while waiting for the remote router to restart
```
Comment (jim-p) [inline net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bgp_neighbors.xml:391]: Reword to avoid the passive voice ("it is wished to ignore").
Should "BD" be "BFD"?

Could be shortened to "This is the case when graceful restart is enabled and BFD should ignore the event while waiting for the remote to restart".

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bgp_neighbors.xml (hunk 13 above)
Comment (jim-p) [inline net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bgp_neighbors.xml:390]: "when a BFD down event is caught" (by what?). Perhaps "When BFD catches a down event".

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_ospf6.xml [hunk 15]
```diff
@@ -131,6 +117,36 @@
 			</description>
 			<type>input</type>
 		</field>
+		<field>
+			<fielddescr>Enable SNMP AgentX</fielddescr>
+			<fieldname>enablesnmpagentx</fieldname>
+			<description>Enable agentx support for accessing FRR Zebra data via SNMP with the net-snmp package.</description>
```
Comment (jim-p) [inline net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_ospf6.xml:123]: Should this be "OSPF6" and not "Zebra"?

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_ospf6_interfaces.xml [hunk 16]
```diff
@@ -184,6 +188,22 @@
...
+			<name>BFD</name>
+			<type>listtopic</type>
+		</field>
+		<field>
+			<fielddescr>BFD Enable</fielddescr>
+			<fieldname>bfdenable</fieldname>
+			<description>
+				<![CDATA[
+				Listen for BFD events on peers created on the interface.
+				Every time a new neighbor is found a BFD peer is created to monitor the link status for fast convergence.
```
Comment (jim-p) [inline net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_ospf6_interfaces.xml:202]: Reword this a bit. Perhaps "When OSPF6 finds a new neighbor, it creates a BFD peer to monitor..."

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_ospf_interfaces.xml [hunk 17]
```diff
@@ -237,6 +237,22 @@
...
+			<name>BFD</name>
+			<type>listtopic</type>
+		</field>
+		<field>
+			<fielddescr>BFD Enable</fielddescr>
+			<fieldname>bfdenable</fieldname>
+			<description>
+				<![CDATA[
+				Listen for BFD events on peers created on the interface.
+				Every time a new neighbor is found a BFD peer is created to monitor the link status for fast convergence.
```
Comment (jim-p) [inline net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_ospf_interfaces.xml:251]: Reword similar to the same block in OSPF6

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr/inc/frr_bfd.inc [hunk 18]
```diff
@@ -19,51 +19,110 @@
  * limitations under the License.
  */
 
+/* Get a list of configured BFD peers */
+function frr_get_bfd_peers($includedefault = true) {
+	global $config;
+	$peers = array();
+
+	if ($includedefault) { $peers[] = array('name' => 'Default', 'value' => 'default'); }
```
Comment (jim-p) [inline net/pfSense-pkg-frr/files/usr/local/pkg/frr/inc/frr_bfd.inc:27]: Please use ["K&R, BSD KNF variant style" formatting](https://www.netgate.com/docs/pfsense/development/developer-style-guide.html#indent-style).

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr/inc/frr_bfd.inc [hunk 19]
```diff
@@ -19,51 +19,110 @@
...
+	if (is_array($config['installedpackages']['frrbfdpeers']['config'])) {
+		foreach ($config['installedpackages']['frrbfdpeers']['config'] as $peer) {
+			$peer_name = $peer['peer'];
+			if (!empty(peer['descr'])) { $peer_name .= " - {$peer['descr']}"; }
```
Comment (jim-p) [inline net/pfSense-pkg-frr/files/usr/local/pkg/frr/inc/frr_bfd.inc:31]: Please use ["K&R, BSD KNF variant style" formatting](https://www.netgate.com/docs/pfsense/development/developer-style-guide.html#indent-style).

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr/inc/frr_bfd.inc [hunk 20]
```diff
@@ -19,51 +19,110 @@
...
+
+	return $peers;
+}
+
+/* Get a list of configured interface addresses */
+function frr_get_bfd_source_addresses($includedefault = true) {
+	$addresses = array();
+	$interfaces = frr_get_interfaces();
+
+	if ($includedefault) { $addresses[] = array('name' => 'Default', 'value' => 'default'); }
```
Comment (jim-p) [inline net/pfSense-pkg-frr/files/usr/local/pkg/frr/inc/frr_bfd.inc:48]: Please use ["K&R, BSD KNF variant style" formatting](https://www.netgate.com/docs/pfsense/development/developer-style-guide.html#indent-style).

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr/inc/frr_bgp.inc [hunk 21]
```diff
@@ -609,15 +609,12 @@ function frr_bgp_generate_neighbors() {
...
 
-		if (frr_validate_intrange($neighbor['bfd_multiplier'], 2, 255)) {
-			$nconf .= "  neighbor {$neighbor['peer']} bfd {$neighbor['bfd_multiplier']}";
-			if (frr_validate_intrange($neighbor['bfd_minrx'], 50, 60000)) {
-				$nconf .= " {$neighbor['bfd_minrx']}";
-				if (frr_validate_intrange($neighbor['bfd_mintx'], 50, 60000)) {
-					$nconf .= " {$neighbor['bfd_mintx']}";
-				}
+		if (!empty($neighbor['bfdenable'])) {
+			$bfd_peer = !empty($neighbor['bfd_peer']) && $neighbor['bfd_peer'] <> 'default' ? $neighbor['bfd_peer'] : $bfd_peer = $neighbor['peer'];
```
Comment (jim-p) [inline net/pfSense-pkg-frr/files/usr/local/pkg/frr/inc/frr_bgp.inc:613]: This line is quite long and the ternary operator use is not properly isolated in parenthesis, which will cause an error on PHP 7.4.

Each section of the test must be in its own pair of parenthesis, plus the whole ternary operator and its parameters must also be in a set of parenthesis.

This is a case where it probably would be best to have it as a proper if statement instead of using the ternary operator, since the current format is more confusing and by the time you line wrap it properly, it isn't saving space.

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd_peers.xml (hunk 8 above)
Comment (bmhughes) [reply net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd_peers.xml:188]: This is straight from the FRR docs, I've reworded as best I can.

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd_peers.xml (hunk 9 above)
Comment (bmhughes) [reply net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd_peers.xml:199]: This is straight from the FRR docs, I've reworded as best I can.

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd_peers.xml (hunk 11 above)
Comment (bmhughes) [reply net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd_peers.xml:217]: This is straight from the FRR docs, I've reworded as best I can.

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd_peers.xml (hunk 10 above)
Comment (bmhughes) [reply net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd_peers.xml:206]: This is straight from the FRR docs, I've reworded as best I can.

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bgp_neighbors.xml (hunk 12 above)
Comment (bmhughes) [reply net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bgp_neighbors.xml:387]: This is straight from the FRR docs, I've reworded as best I can.

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bgp_neighbors.xml (hunk 13 above)
Comment (bmhughes) [reply net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bgp_neighbors.xml:390]: This is straight from the FRR docs, I've reworded as best I can.

Comment (bmhughes) [reply net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bgp_neighbors.xml:390]: I presume "stalled" from the context.

**Comment by bmhughes** [comment]
Changes made and rebase on current devel.

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd_peers.xml [hunk 22]
```diff
@@ -0,0 +1,231 @@
...
+		</field>
+
+		<field>
+			<name>Source Address/Interface</name>
+			<type>listtopic</type>
+		</field>
+		<field>
+			<fielddescr>Interface</fielddescr>
+			<fieldname>interface</fieldname>
+			<description>Select the interface to source BFD packets from </description>
```
Comment (jim-p) [inline net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd_peers.xml:143]: There is an extra space at the end of this sentence, also not that it's a huge problem, but ending in a preposition isn't ideal.
"Select the interface which BFD will use as a packet source" is one good way around that.

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd_peers.xml [hunk 23]
```diff
@@ -0,0 +1,231 @@
...
+			<source><![CDATA[frr_get_interfaces()]]></source>
+			<source_name>name</source_name>
+			<source_value>value</source_value>
+		</field>
+		<field>
+			<fielddescr>Local Source Address</fielddescr>
+			<fieldname>local_address</fieldname>
+			<description>
+				<![CDATA[
+				Provide a local address to bind our BFD peer listener to for participating in a BFD session.<br />
```
Comment (jim-p) [inline net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd_peers.xml:154]: "our" is also one you should avoid in this context. The intent is to avoid addressing the user as much as possible (See https://docs.netgate.com/pfsense/en/latest/references/style-guide.html#what-to-avoid ).

In this case you can replace "our" with "the" and it works.

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd_peers.xml (hunk 8 above)
Comment (jim-p) [reply net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd_peers.xml]: Even if it is from the FRR docs, we prefer to match our own style when bringing in text as much as possible.

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd_peers.xml [hunk 24]
```diff
@@ -0,0 +1,231 @@
...
+		<field>
+			<fielddescr>Echo Mode</fielddescr>
+			<fieldname>echo_mode</fieldname>
+			<description>
+				<![CDATA[
+				Enables or disables the echo transmission mode. This mode is disabled by default.
+				<br />
+				FRR documentation recommendeds that the transmission interval of control packets to be increased after enabling echo-mode to reduce bandwidth usage. For example: transmission-interval 2000.
+				<br />
+    			Echo mode is not supported on multi-hop setups, see RFC 5883 section 3.
```
Comment (jim-p) [inline net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd_peers.xml:219]: This line needs its indent level fixed to match its surrounding lines

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd_peers.xml (hunk 22 above)
Comment (bmhughes) [reply net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bfd_peers.xml:143]: No problem, done.

**Comment by bmhughes** [comment]
Revised with the updated changes.

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bgp_neighbors.xml [hunk 25]
```diff
@@ -375,6 +375,47 @@
...
+		</field>
+		<field>
+			<fielddescr></fielddescr>
+			<description>
+				<![CDATA[
+				<b>BFD Enable</b>: Listen for BFD events registered on the same target as this BGP neighbor.
+				<br />
+				<b>BFD Check Control Plane Failure</b>: Allow to write CBIT independence in BFD outgoing packets. Also allow to read both C-BIT value of BFD and lookup BGP peer status.
+				This command is useful when BFD catches a down event is caught, while the BGP peer requested that local BGP keeps the remote BGP entries as stalled if such issue is detected.
+				This is the case when graceful restart is enabled and BFD should ignore the event while waiting for the remote to restart.
```
Comment (jim-p) [inline net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_bgp_neighbors.xml:390]: Looking at RFC 4724 it's probably meant to be "stale".

New text should be:

> Allow FRR to write CBIT independence in outgoing BFD packets. Also allow FRR to read both the CBIT value of BFD and lookup BGP peer status.
> This option allows BFD to ignore down events during a graceful restart of the remote peer if graceful restarts are enabled in BGP.
> When enabled, if BFD catches a down event it first checks if the BGP peer has requested that local the BGP daemon keep the remote BGP entries marked as stale. In that case it can safely ignore the event to allow the restart to happen gracefully (RFC 4724).

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr/inc/frr_bgp.inc [hunk 26]
```diff
@@ -609,15 +609,16 @@ function frr_bgp_generate_neighbors() {
...
 
-		if (frr_validate_intrange($neighbor['bfd_multiplier'], 2, 255)) {
-			$nconf .= "  neighbor {$neighbor['peer']} bfd {$neighbor['bfd_multiplier']}";
-			if (frr_validate_intrange($neighbor['bfd_minrx'], 50, 60000)) {
-				$nconf .= " {$neighbor['bfd_minrx']}";
-				if (frr_validate_intrange($neighbor['bfd_mintx'], 50, 60000)) {
-					$nconf .= " {$neighbor['bfd_mintx']}";
-				}
+		if (!empty($neighbor['bfdenable'])) {
+			if (!empty($neighbor['bfd_peer']) && $neighbor['bfd_peer'] <> 'default') {
```
Comment (jim-p) [inline net/pfSense-pkg-frr/files/usr/local/pkg/frr/inc/frr_bgp.inc:613]: The second test here `$neighbor['bfd_peer'] <> 'default'` should be in its own set of parenthesis.
